misc-scripts/canonical_transcripts/select_canonical_transcripts.pl

```diff
@@ -248,10 +248,36 @@
     my $gene_update_sql = "UPDATE gene SET canonical_transcript_id = ? where gene_id = ?";
     my $gene_sth = $dba->dbc->prepare($gene_update_sql);
 
+    # Prepare transcript canonical attribute sqls
+    my $trans_select_sth = $dba->dbc->prepare("SELECT value FROM transcript_attrib WHERE transcript_id=? AND attrib_type_id=?");
+    my $trans_update_sth = $dba->dbc->prepare("UPDATE transcript_attrib SET value=? WHERE transcript_id=? AND attrib_type_id=?");
+    my $trans_delete_sth = $dba->dbc->prepare("DELETE FROM transcript_attrib WHERE transcript_id=? AND attrib_type_id=?");
+
+    # Get attribute id for canonical
+    my $attrib_type_id_sth = $dba->dbc->prepare("SELECT attrib_type_id FROM attrib_type WHERE code=?");
+    $attrib_type_id_sth->execute('is_canonical');
+    my ($attrib_type_id) = $attrib_type_id_sth->fetchrow_array() || 554;
```

Comment:
attrib_type_id will not change within the same database, so this should be moved outside the foreach my slice loop. (In theory, it should be consistent across all databases because it's synchronised to `master_attrib_type` in the `ensembl_production` db). It also should not default to 554. If, for some reason, there is no `is_canonical` attrib in that databases `attrib_type` table, then I think the script should fail, rather than insert attribs with a type that isn't in `attrib_type`. This also guards against the `attrib_type_id` changing if for some reason Production has to change it.

Reply:
Having discussed this with Stefano, he does agree with you Brandon that the script should fail if for some reason we can't find an id for the is_canonical attribute. I was trying to make sure the script still works regardless of my changes (make it backwards compatible and all), but I see now that that isn't an issue. I have changed the code to reflect that.